Lib/test/test_shelve.py

```diff
@@ -165,6 +166,277 @@ def test_default_protocol(self):
+                return data.decode("utf-8")
+            elif isinstance(data, array.array):
+                return array.array("b", data)
+            else:
+                raise TypeError(
+                    f"Unsupported type for deserialization: {type(data)}"
+                )
+
+        os.mkdir(self.dirname)
+        self.addCleanup(os_helper.rmtree, self.dirname)
+
+        for proto in range(pickle.HIGHEST_PROTOCOL + 1):
+            with self.subTest(proto=proto):
+                with shelve.open(
+                    self.fn,
+                    protocol=proto,
+                    serializer=serializer,
+                    deserializer=deserializer,
+                ) as s:
+                    bar = "bar"
+                    bytes_data = b"Hello, world!"
+                    bytearray_data = bytearray(b"\x00\x01\x02\x03\x04")
+                    array_data = array.array("i", [1, 2, 3, 4, 5])
+
+                    s["foo"] = bar
+                    s["bytes_data"] = bytes_data
+                    s["bytearray_data"] = bytearray_data
+                    s["array_data"] = array_data
+
+                    if proto == 5:
+                        self.assertEqual(s["foo"], str(bar))
+                        self.assertEqual(s["bytes_data"], "Hello, world!")
+                        self.assertEqual(
+                            s["bytearray_data"], bytearray_data.decode()
+                        )
+                        self.assertEqual(
+                            s["array_data"], array_data.tobytes().decode()
+                        )
+                    else:
+                        self.assertEqual(s["foo"], "str")
+                        self.assertEqual(s["bytes_data"], "bytes")
+                        self.assertEqual(s["bytearray_data"], "bytearray")
+                        self.assertEqual(
+                            s["array_data"], array_data.tobytes().decode()
+                        )
+
+    def test_custom_incomplete_serializer_and_deserializer(self):
+        dbm_sqlite3 = import_helper.import_module("dbm.sqlite3")
+        os.mkdir(self.dirname)
+        self.addCleanup(os_helper.rmtree, self.dirname)
+
+        with self.assertRaises(dbm_sqlite3.error):
+            def serializer(obj, protocol=None):
+                pass
+
+            def deserializer(data):
+                return data.decode("utf-8")
+
+            with shelve.open(self.fn, serializer=serializer,
+                             deserializer=deserializer) as s:
+                s["foo"] = "bar"
+
+        def serializer(obj, protocol=None):
+            return type(obj).__name__.encode("utf-8")
+
+        def deserializer(data):
+            pass
+
+        with shelve.open(self.fn, serializer=serializer,
+                         deserializer=deserializer) as s:
+            s["foo"] = "bar"
+            self.assertNotEqual(s["foo"], "bar")
+            self.assertEqual(s["foo"], None)
```

Comment:
```suggestion
            self.assertIsNone(s["foo"])
```

I'll let you change the other occurrences (if any).